syntax = "proto3";

package healthservice;

// This is interface exported by the server.

// define request and response structure for rpc calls

// Defines the User/Researcher data access request grpc communicate data
// structure

// These messages deal with authenticating a requester.
message RequestUserRecordAccessResearcherToUser {
  string patient_blockchain_address = 1;
}
message ResponseUserRecordAccessUserToResearcher {
  string patient_blcokchain_address = 1;
  string signed_data_request = 2; // signed with researcher and user keys
}
message RequestUserRecordAccessResearcherToVO {
  string patient_blockchain_address = 1;
  string signed_data_request = 2;
  string jwt_token = 3;
}
message ResponseUserRecordAccessVOToResearcher {
  string patient_blockchain_address = 1;
  // data chunk sent back to chain. Need to confirm exactly what's being sent
  // across the wire
  string signed_data_request = 2;
}

// TODO: Confirm with Graham that this is the correct structure
message SendUserAuthTokenToResearcher {
  string patient_blockchain_address = 1;
  string jwt_token = 2;
}
message SendUserAuthTokenToVO {
  string patient_blockchain_address = 1;
  string jwt_token = 2;
}

message ConfirmUserAuthTokenVOToUser { bool success = 1; }
message ConfirmUserAuthTokenResearcherToUser { bool success = 1; }

// These messages deal with the handshake process and the gossiping of
// transactions and blocks
message HandshakeRequest {
  string version = 1;
  string time = 2;
  string addrMe = 3;
  int32 bestHeight = 4;
}

message HandshakeResponse { repeated string knownPeers = 1; }

message NewTransactionRequest { 
  Transaction newTransaction = 1; 
}

message Transaction { 
  int32 VersionNumber = 1; 
  string Patient_Ad = 2; 
  string VOAddress = 3; 
  int32 HippaID = 4; 
  bool SummaryAvail = 5; 
  string Data = 6; 
  string RequestorAddress = 7; 
  string Approval = 8; 
  string TransactionHash = 9; 
  patientDB patient_db = 10; 
}

message patientDB {
    repeated patient patients = 1; 
}

message patient {
    string first_name = 1;
    string last_name = 2;
    string email = 3;
}

message NewBlockRequest { Block newBlock = 1; }

message Block {
    string MagicNumber = 1; 
    int32 Blocksize = 2; 
    repeated Transaction Transactions_input = 3; 
    repeated string Transactions = 4; 
    int32 TransactionCounter = 5; 
    Header BlockHeader = 6; 
    string Blockhash = 7; 
}

message Header {
    repeated string Block_Transactions = 1; 
    int32 Version = 2; 
    string hashPrevBlock = 3; 
    repeated string MerkleTree = 4; 
    int64 Timestamp = 5; 
    int32 Bits = 6; 
    int32 Nonce = 7; 
}

// These messages deal with the retrieval of health records

// Single health record
message HealthRecordResponse {
  int32 id = 1;
  string first_name = 2;
  string last_name = 3;
  string email = 4;
  string gender = 5;
  float height = 6;
  float weight = 7;
  int32 age = 8;
  int32 HippaID = 9;
  string patient_address = 10;
  string vo_address = 11;
}
message HealthRecordListResponse {
  repeated HealthRecordResponse health_records = 1;
}

// TODO What do we send to the VO as a verified requester?
message GetHealthRecordsRequest { Empty request = 1; }

message Empty {}

service HealthService {

  rpc getUserRecordAccessResearcherToUser(
      RequestUserRecordAccessResearcherToUser)
      returns (ResponseUserRecordAccessUserToResearcher);

  rpc getUserRecordAccessResearcherToVO(RequestUserRecordAccessResearcherToVO)
      returns (ResponseUserRecordAccessVOToResearcher);

  rpc sendUserAuthTokenToResearcher(SendUserAuthTokenToResearcher)
      returns (ConfirmUserAuthTokenResearcherToUser);
  rpc sendUserAuthTokenToVO(SendUserAuthTokenToVO)
      returns (ConfirmUserAuthTokenVOToUser);

  rpc getHealthRecords(GetHealthRecordsRequest)
      returns (HealthRecordListResponse);

  rpc Handshake(HandshakeRequest) returns (HandshakeResponse);
  rpc NewTransactionBroadcast(NewTransactionRequest) returns (Empty);
  rpc NewBlockBroadcast(NewBlockRequest) returns (Empty);
}
